capa/features/extractors/binexport2/insn.py

```diff
@@ -0,0 +1,392 @@
+
+            if symbol.startswith("#0x"):
+                # like:
+                # - type: SYMBOL
+                #   symbol: "#0xffffffff"
+                try:
+                    value = int(symbol[len("#") :], 0x10)
+                except ValueError:
+                    # failed to parse as integer
+                    continue
+
+                # handling continues below at label: has a value
+
+            elif symbol.startswith("0x"):
+                # like:
+                # - type: SYMBOL
+                #   symbol: "0x1000"
+                try:
+                    value = int(symbol, 0x10)
+                except ValueError:
+                    # failed to parse as integer
+                    continue
+
+                # handling continues below at label: has a value
+
+            else:
+                continue
+
+        elif len(operand.expression_index) == 2:
+            # from BinDetego,
+            # we get the following pattern for immediate constants:
+            #
+            # - type: SIZE_PREFIX
+            #   symbol: "b8"
+            # - type: IMMEDIATE_INT
+            #   immediate: 20588728364
+            #   parent_index: 0
+
+            expression0 = be2.expression[operand.expression_index[0]]
+            expression1 = be2.expression[operand.expression_index[1]]
+
+            if BinExport2.Expression.Type.SIZE_PREFIX != expression0.type:
+                continue
+
+            if BinExport2.Expression.Type.IMMEDIATE_INT != expression1.type:
+                continue
+
+            value = expression1.immediate
+
+            # handling continues below at label: has a value
+
+        else:
+            continue
+
+        # label: has a value
+
+        if analysis.base_address == 0x0:
+            # When the image is mapped at 0x0,
+            #  then its hard to tell if numbers are pointers or numbers.
+            # TODO(mr): 1755 be a little less conservative otherwise?
+
+            # TODO(mr): 1755 this removes a lot of valid numbers, could check alignment and use additional heuristics
+            # if is_address_mapped(be2, value):
+            #     continue
+            pass
+
+        yield Number(value), ih.address
+        yield OperandNumber(i, value), ih.address
+
+
+class ReadMemoryError(ValueError): ...
+
+
+def read_memory(ctx: AnalysisContext, sample_bytes: bytes, address: int, size: int, cache) -> bytes:
+    rva = address - ctx.analysis.base_address
+
+    try:
+        if sample_bytes.startswith(capa.features.extractors.common.MATCH_PE):
+            pe = cache.get("pe")
+            if not pe:
+                pe = pefile.PE(data=sample_bytes)
+                cache["pe"] = pe
+            return pe.get_data(rva, size)
+        elif sample_bytes.startswith(capa.features.extractors.common.MATCH_ELF):
+            elf = cache.get("elf")
+            if not elf:
+                elf = ELFFile(io.BytesIO(sample_bytes))
+                cache["elf"] = elf
```

Comment:
It appears that we're creating a new `PE`/`ELFFILE` object per function, when we should be creating these once per analysis. Commenting out `extract_insn_bytes_features` (the caller) resulted in a significant performance improvement. @mr-tz this is a good place to start investigating for a performance bump, starting by seeing if we can store the `PE`/`ELFFile` object once in the `AnalysisContext`.

Reply:
My guess is that each read call is pretty expensive because it parses through all the headers and sections to find the right region to slice, every time.  It's naive but simple to read. 

We could preprocess the PE/ELF into a list of `(start, end, bytes)` and read from there. I bet this would be a lot faster at the expense of a little memory (~input file size, assuming we don't NULL extend regions).

Reply:
See Willi's response, I think this part causes the slowness:

https://github.com/mandiant/capa/pull/1950/files#diff-43e6278fbf1b9545b05273c6a12f915da9938933c68022bd9dcb11f284d2ad74R229-R244

Reply:
In FLOSS we cache using memoryview, see e.g. https://github.com/mandiant/flare-floss/blob/bcaf4706b585501ae980dbce0a042e03550c29f4/floss/language/utils.py#L395-L408

Reply:
I've temp disabled this in [1fea6ab](https://github.com/mandiant/capa/pull/1950/commits/1fea6ab5cee4edf21b7250e3aaa25b902a263498) until we can improve its performance. While testing locally the capa process is getting killed for larger samples so we need to remove this blocker for now to continue development and rule writing.

Reply:
can you share a sample with poor performance that i can use to guide optimization?

Reply:
Absolutely, [mimikatz.exe_ from our CAPA test files](https://github.com/mandiant/capa-testfiles/blob/master/mimikatz.exe_). While testing locally with this code enabled the CAPA process is killed when running against mimikatz.exe_. CAPA runs as expected when this code is disabled. Unfortunately I haven't had time to dig in further than that although the point of process kill appears nondeterministic.

Reply:
if it's easy, do you have a binexport for that file handy? if not, not a big deal.

Reply:
work tracked in https://github.com/mandiant/capa/issues/2061

Reply:
fixed in 87604811cbc3eea41a6aa94c85607b6a4ff3c135